nixos/networkd: fix wait-online on networking.useDHCP default configuration

lf- opened the pull request:
## Description of changes

Due to this being fixed: https://github.com/systemd/systemd/issues/27822 systemd-networkd will now wait for interfaces to appear that are WaitForOnline. However, if there are no such interfaces, it times out! Oops!

I filed it upstream, and the systemd developers pointed to the previous issue and questioned why we were turning that unit on to begin with. I think they're probably right here: the correct move is *probably* to turn off the unit if there is nothing configured. https://github.com/systemd/systemd/issues/29388

Regarding testing, I honestly don't even know how to begin to write a reliable test for this. I have poked at it in the repl a little bit and verified it appears to evaluate to the expected values.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

lf- commented:
I *think* the right people to ping here are @NixOS/systemd

ElvishJerricco commented:
This doesn't seem right to me. We want the service *to exist*, but it shouldn't be a dependency of `multi-user.target` by default. From `man systemd.special` about `network-online.target`:

> Note the distinction between this unit and network.target. This unit is an active unit (i.e. pulled in by the consumer rather than the provider of this functionality) and pulls in a service which possibly adds substantial delays to further execution. In contrast, network.target is a passive unit (i.e. pulled in by the provider of the functionality, rather than the consumer) that usually does not delay execution much. Usually, network.target is part of the boot of most systems, while network-online.target is not, except when at least one unit requires it.

Services that need the network to be online should pull it in, but that certainly does not include `multi-user.target`, which suggest that this line is what's wrong:

https://github.com/NixOS/nixpkgs/blob/dd8af63ff7683947008161126020971ddc5091d6/nixos/modules/system/boot/systemd.nix#L614

flokli commented:
Yes, if you check the history of that file, you see this mostly got reintroduced because too many tests waiting for network to be up actually only wait for multi-user.target.

https://github.com/NixOS/nixpkgs/commit/15d761a525a025de0680b62e8ab79a9d183f313d#commitcomment-128564097

We should probably remove it from there again, and re-add it to every VM test relying on that (or better, update the VM test to wait on the right target(s).

Cc @arianvp

lf- commented:
The suggestion that the systemd devs gave was to set `anyInterface` to true by default then stop setting `RequiredForOnline=no` in our default network configs. This *also* seems like a reasonable implementation, probably a less fucked up one than mine.

ElvishJerricco commented:
Oh but I see, the additional issue is that with our very broad `networking.useDHCP = true` network configuration, it's *going* to time out if there are multiple interfaces and not all are connected (which is quite common). So yes, I agree that we should:

1. Remove the dependency from `multi-user.target` (a *very crude* stopgap is to add the dependency back in `test-instrumentation.nix` so that all VMs have it again; and we can start fixing them all in a followup PR).
2. Enable `anyInterface` by default when using `networking.useDHCP` and remove the `RequireForOnline=no` bit.

lf- commented:
Can we somehow roll an epoch for tests? Like, do this janky workaround and insert a "stateVersion" equivalent thing into all the existing tests and then apply the new behaviour by default?

flokli commented:
> Can we somehow roll an epoch for tests? Like, do this janky workaround and insert a "stateVersion" equivalent thing into all the existing tests and then apply the new behaviour by default?

I don't see why this is necessary. stateVersion is meant to deal with changes of *state*, and test VMs are ephemeral.

I'm fine with adding this hack into test-instrumentation.nix and subsequently updating VM tests to work without this assumption, but I don't think we should add brittle code trying to solve a problem we shouldn't have in first place.

lf- commented:
Incidentally I found out on mastodon that our network-manager integration contains what might be a very similar bug: https://github.com/NixOS/nixpkgs/issues/180175.

I'm going to rewrite this PR to follow the plan set out by @ElvishJerricco.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/why-does-multi-user-target-depend-on-network-online-target/33565/10

lf- committed:
nixos/networkd: make wait-online not time out

A further bug to our strange multi-user.target depending on
network-online.target issue is that systemd recently changed the
behaviour of systemd-networkd-wait-online to no longer consider the
absence of interfaces with RequiredForOnline to be sufficient to be
online: https://github.com/systemd/systemd/pull/27825

On the advice of the systemd developers
(https://github.com/systemd/systemd/issues/29388), this commit changes
the configuration of systemd-networkd-wait-online to pass --any by
default, and lets the default DHCP interfaces be RequiredForOnline
as they would be by default if the option is omitted.

It is plausible that systemd-networkd-wait-online may still fail if
there are no interfaces at all. However, that probably cannot be
avoided.

lf- commented:
I have rewritten the PR to follow with the advice of the systemd devs (and also rewritten the commit messages). I split the two changes into separate commits in case they need to be reverted.

lf- commented:
I have locally run nixosTests.networking.networkd and they all pass, so, I *think* it works?

ElvishJerricco commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -59,8 +59,12 @@ let
+      # systemd no longer considering "all interfaces not required for online"
+      # to be sufficient for being online
+      # <https://github.com/systemd/systemd/issues/29388>), RequiredForOnline
+      # really means "sufficient for online", so we can enable it.
```
I thought the idea was that we should *remove* `RequiredForOnline=no` and instead just leave it unconfigured and let `--any` handle this?

lf- replied:
Leaving it as unconfigured is equivalent to setting it to true with the exception of if it is set to down by default. https://github.com/systemd/systemd/blob/117843fe95f091df1ddc87aaa60f1a2a8698d120/src/network/networkd-network.c#L282-L292

My thought process with leaving this code here is that if you *are* manually setting up networks *and* deliberately unset --any, you definitely don't want the default networks to be required to be considered online (e.g. in the case of a laptop with an Ethernet card that is not plugged in). But then again in that case I feel like you just want to turn off the default networks.

Feels like the global setting of --any is a clumsy tool from upstream, and I don't really know how to effectively use it.

ElvishJerricco reviewed (commented):
I'm a little confused here, so I'll sum up what I mentioned in the comments that confused me. When using `newtorking.useDHCP`, the problem was that we didn't want it to block waiting for *every* generically named interface that matches our `genericDhcpNetworks`, as it's quite common for a device to have many generic interfaces with only one plugged in. Previously, we used `RequiredForOnline = false;` so that `systemd-networkd-wait-online` wouldn't have to wait for all those interfaces. This was wrong because it caused `systemd-networkd-wait-online` to do nothing useful whatsoever.

The new behavior in 254 is no longer allowing us to make that mistake. The systemd folks suggested that the behavior we actually want is the behavior of `--any`, not the old `RequiredForOnline=no` behavior. This way `systemd-networkd-wait-online` does actually wait for at least one interface to be configured, which is useful. But it *doesn't* require *all* of them to be configured, eliminating the need for the `RequiredForOnline` trick we used before that no longer works. And importantly, we only want any of this in the first place if we have `networking.useDHCP`; we don't want it when the user has explicit network configuration instead of `networking.useDHCP` because that indicates that they do actually want all those interfaces to become configured.

So we should remove the `RequiredForOnline` trick, and we should set `anyInterface` so that it's enabled by default only if `networking.useDHCP` is enabled; otherwise enabling `anyInterface` is *not* a correct default, as explicit network configuration is in effect, so proper waiting is desired.

Does that sound like the behavior we want?

lf- commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -2724,9 +2724,13 @@ let
         '';
         type = types.bool;
-        default = false;
+        default = true;
```
Hm, I can see this reducing the any compatibility break.

lf- commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -2761,6 +2765,24 @@ let
+      isRequiredForOnline = net: !(elem (net.linkConfig.RequiredForOnline or false) [false "no"]);
+    in (mkIf cfg.enable {
+      assertions = [{
+        assertion = cfg.wait-online.enable -> (!cfg.wait-online.anyInterface -> any id (mapAttrsToList (_k: isRequiredForOnline) cfg.networks));
```
removed

lf- commented:
> Does that sound like the behavior we want

I think so. We can use the default behaviour of RequiredForOnline = true plus setting anyInterface in the useDHCP case, and let anyInterface otherwise default to off.

ElvishJerricco commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -91,6 +91,10 @@
 
+- `network-online.target` is no longer required by `multi-user.target`. This improves overall boot times and is consistent with the behaviour of other distributions.
+
+  Relatedly, `network-online.target` has been fixed to no longer time out for systems with `networking.useDHCP = true` and `systemd.network.enable = true`.
```
Should probably say `networking.useNetworkd` instead of `systemd.network.enable`, since the generic network is only configured when the former is enabled.

ElvishJerricco reviewed (changes requested):
Other than this last little nitpick, LGTM

ElvishJerricco commented:
@ofborg eval

rnhmjoj commented on nixos/modules/testing/test-instrumentation.nix:
```diff
@@ -131,6 +131,19 @@ in
+    # - https://github.com/systemd/systemd/issues/29388
+    # - https://github.com/NixOS/nixpkgs/issues/180175
+    # - https://github.com/NixOS/nixpkgs/commit/15d761a525a025de0680b62e8ab79a9d183f313d#commitcomment-128564097
+    systemd.targets.network-online.wantedBy = [ "multi-user.target" ];
```
Can we fix these tests properly? I think this hack has been going on for too long already.

rnhmjoj replied:
Do I understand correctly that this is just about replacing a bunch of `wait_for_unit("multi-user.target")` with `wait_for_unit("network-online.target")`?

lf- replied:
Probably. I don't have the machine time to go and run the entire testsuite and find out which ones are sketchy. It's also possible that some services themselves have their deps mis-declared. Overall this is a big can o worms.

I'm actually feeling quite dubious about the idea of patching the tests with a bad hack, because it means that the tests won't catch the bugs, our users perhaps will.

rnhmjoj replied:
Exactly: https://github.com/NixOS/nixpkgs/pull/86273 mentions services breaking, so this PR is already likely to cause a breakage already. Artificially keeping the tests up but breaking the services may be even worse.

lf- replied:
Hmmm. I guess we will have to figure out who to poke to get some Hydra resources to run all the tests. Actually this is silly. Let's split the dependency change part off this PR, ship the bug fix, and make a new PR to explicitly work through the fallout of the dependency change. I will try to get that done tomorrow.

ElvishJerricco replied:
@rnhmjoj I don't think we should block this PR on fixing the tests. That could be a significant effort, and it's not relevant to what this PR is ultimate trying to do. I'd rather get this long-needed bugfix in without having to stall on tests that are already working ok (if a bit wastefully)

lf- replied:
I think we should leave in the false dependency here because it's not explicitly breaking anything and fix the false dependency by itself in a second pr.

rnhmjoj replied:
Agreed with everything said.

lf- commented:
Alright I have pulled out the `multi-user.target` fix commit and moved it to here: https://github.com/NixOS/nixpkgs/pull/258680

*This* PR can then be merged to fix networkd, leaving the false dependency intact, and then we can work on fixing up the false dependency separately.
